File: day08/src/main.py
from collections import defaultdict
from pathlib import Path


def parse_input(file_path):
    data = Path(file_path).read_text().strip().split('\n')
    antennas = [(x, y, freq) for y, line in enumerate(data) for x, freq in enumerate(line) if freq != '.']
    width, height = len(data[0]), len(data)
    return antennas, width, height


def find_antinodes(antennas, width, height, part_two=False):
    antinodes = set()
    frequency_map = defaultdict(list)

    for x, y, freq in antennas:
        frequency_map[freq].append((x, y))

    for freq, positions in frequency_map.items():
        n = len(positions)

        for i in range(n):
            x1, y1 = positions[i]

            for j in range(i + 1, n):
                x2, y2 = positions[j]
                dx, dy = x2 - x1, y2 - y1

                for heading, start_x, start_y in (-1, x1, y1), (1, x2, y2):
                    step = 0 if part_two else heading

                    while True:
                        xa, ya = start_x + step * dx, start_y + step * dy

                        if not (0 <= xa < width and 0 <= ya < height):
                            break

                        antinodes.add((xa, ya))

                        if not part_two:
                            break

                        step += heading

    return antinodes


def part1():
    antennas, width, height = parse_input('my_input.txt')
    antinodes = find_antinodes(antennas, width, height, part_two=False)
    print(f'Answer: {len(antinodes)}')


def part2():
    antennas, width, height = parse_input('my_input.txt')
    antinodes = find_antinodes(antennas, width, height, part_two=True)
    print(f'Answer: {len(antinodes)}')


if __name__ == '__main__':
    part1()
    part2()

File: day08/src/main.rs
use std::collections::{HashMap, HashSet};
use std::fs;

#[derive(Debug)]
struct Antenna {
    x: isize,
    y: isize,
    freq: char,
}

#[derive(Debug)]
struct Grid {
    antennas: Vec<Antenna>,
    width: isize,
    height: isize,
}

fn parse_input(filename: &str) -> Grid {
    let data = fs::read_to_string(filename).expect("Failed to read file");
    let lines: Vec<_> = data.trim().lines().collect();

    Grid {
        antennas: lines
            .iter()
            .enumerate()
            .flat_map(|(y, line)| {
                line.trim()
                    .chars()
                    .enumerate()
                    .filter(|(_, freq)| *freq != '.')
                    .map(move |(x, freq)| Antenna {
                        x: x as isize,
                        y: y as isize,
                        freq,
                    })
            })
            .collect(),
        width: lines.first().map_or(0, |line| line.trim().len() as isize),
        height: lines.len() as isize,
    }
}

fn find_antinodes(grid: &Grid, is_part_two: bool) -> HashSet<(isize, isize)> {
    let mut antinodes = HashSet::new();

    let mut frequency_map: HashMap<char, Vec<&Antenna>> = HashMap::new();
    for antenna in &grid.antennas {
        frequency_map.entry(antenna.freq).or_default().push(antenna);
    }

    for positions in frequency_map.values() {
        for (i, a1) in positions.iter().enumerate() {
            for a2 in positions.iter().skip(i + 1) {
                let dx = a2.x - a1.x;
                let dy = a2.y - a1.y;

                for (heading, start_x, start_y) in [(-1, a1.x, a1.y), (1, a2.x, a2.y)] {
                    let mut step = if is_part_two { 0 } else { heading };

                    loop {
                        let xa = start_x + step * dx;
                        let ya = start_y + step * dy;

                        if xa < 0 || xa >= grid.width || ya < 0 || ya >= grid.height {
                            break;
                        }

                        antinodes.insert((xa, ya));

                        if !is_part_two {
                            break;
                        }

                        step += heading;
                    }
                }
            }
        }
    }

    antinodes
}

fn part1() {
    let grid = parse_input("my_input.txt");
    let antinodes = find_antinodes(&grid, false);
    println!("Answer: {}", antinodes.len());
}

fn part2() {
    let grid = parse_input("my_input.txt");
    let antinodes = find_antinodes(&grid, true);
    println!("Answer: {}", antinodes.len());
}

fn main() {
    part1();
    part2();
}
